contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+{
+
+}
+
+void USDScene::writeTags( const SceneInterface::NameList &tags )
+{
+	throw IECore::NotImplementedException("USDScene::writeTags not supported" );
+}
+
+bool USDScene::hasObject() const
+{
+	return isConvertable( m_prim );
+}
+
+PrimitiveVariableMap USDScene::readObjectPrimitiveVariables( const std::vector<InternedString> &primVarNames, double time ) const
+{
+	return PrimitiveVariableMap();
+}
+
+void USDScene::writeObject( const Object *object, double time )
+{
+	throw IECore::NotImplementedException("USDScene::writeObject not supported" );
+}
+
+bool USDScene::hasChild( const SceneInterface::Name &name ) const
+{
+	SceneInterface::NameList children;
+	childNames( children );
+	for( const auto &i : children )
+	{
+		if( i == name )
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+void USDScene::childNames( SceneInterface::NameList &childNames ) const
+{
+	for( const auto &i : m_prim.GetAllChildren() )
+	{
+		if( i.GetTypeName() == "Xform" || isConvertable( i ) )
+		{
+			childNames.push_back( IECore::InternedString( i.GetName() ) );
+		}
+	}
+}
+
+SceneInterfacePtr USDScene::child( const SceneInterface::Name &name, SceneInterface::MissingBehaviour missingBehaviour )
+{
+	for( const auto &i : m_prim.GetAllChildren() )
+	{
+		if( ( i.GetTypeName() == "Xform" || isConvertable( i ) ) && name == SceneInterface::Name( i.GetName() ) )
+		{
+			SceneInterfacePtr newScene = new USDScene( m_root, i );
+			return newScene;
+		}
+	}
+
+	switch( missingBehaviour )
+	{
+		case SceneInterface::NullIfMissing :
+			return nullptr;
+		case SceneInterface::ThrowIfMissing :
+			throw IOException( "Child \"" + name.string() + "\" does not exist" );
+		case SceneInterface::CreateIfMissing :
+			throw InvalidArgumentException( "Child creation not supported" );
+		default:
+			return nullptr;
+	}
+}
+
+ConstSceneInterfacePtr USDScene::child( const SceneInterface::Name &name, SceneInterface::MissingBehaviour missingBehaviour ) const
+{
+	return const_cast<USDScene *>( this )->child( name, missingBehaviour );
+}
+
+SceneInterfacePtr USDScene::createChild( const SceneInterface::Name &name )
+{
+	throw IECore::NotImplementedException("USDScene::createChild not supported" );
+	return nullptr;
+}
+
+SceneInterfacePtr USDScene::scene( const SceneInterface::Path &path, SceneInterface::MissingBehaviour missingBehaviour )
+{
+	pxr::UsdPrim prim = m_root->root();
+
+	for( const Name &name : path )
+	{
+		prim = prim.GetChild( pxr::TfToken( name ) );
+	}
+	return new USDScene( m_root, prim );
+}
+
+ConstSceneInterfacePtr USDScene::scene( const SceneInterface::Path &path, SceneInterface::MissingBehaviour missingBehaviour ) const
+{
+	return const_cast<USDScene *>( this )->scene( path, missingBehaviour );
+}
+
+void USDScene::hash( SceneInterface::HashType hashType, double time, MurmurHash &h ) const
+{
+	SceneInterface::hash( hashType, time, h );
+
+	h.append( hashType );
+
+	auto transformHash = [this]( double time, MurmurHash &h )
```

Comment:
ignore the use of lambdas here i'm just mucking around.

Reply:
I think this is interesting actually. Lucio didn't want to have separate `hash*()` etc methods to match the existing `read*()` methods so he went for the `hash( what )` approach we have now. I never liked that because I don't think it actually helps the caller, it's different to the rest of the design, and it results in the implementations being a mess internally. Maybe we should just split it into separate virtual methods now we have the chance?

Reply:
What I meant by ignore the lambdas is I'll change this to member functions like AlembicScene.  I prefer the separate hash functions and we should refactor SceneInterface now. I'll put this on my todo list if you don't mind?

Reply:
> I prefer the separate hash functions and we should refactor SceneInterface now. I'll put this on my todo list if you don't mind?

So you'll make non-virtual member functions internal to USDScene for now, and then do the refactor later? Sounds good.